qiskit/pulse/ir/ir.py

```diff
@@ -0,0 +1,375 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=cyclic-import
+
+"""
+=========
+Pulse IR
+=========
+
+"""
+
+from __future__ import annotations
+from typing import List, Callable
+from abc import ABC, abstractmethod
+
+import numpy as np
+
+from qiskit.pulse.exceptions import PulseError
+
+from qiskit.pulse.transforms import AlignmentKind, AlignSequential
+from qiskit.pulse.instructions import Instruction
+
+
+class IrElement(ABC):
+    """Base class for Pulse IR elements"""
+
+    @property
+    @abstractmethod
+    def initial_time(self) -> int | None:
+        """Return the initial time of the element"""
+        pass
+
+    @property
+    @abstractmethod
+    def duration(self) -> int | None:
+        """Return the duration of the element"""
+        pass
+
+    @abstractmethod
+    def shift_initial_time(self, value: int):
+        """Shift ``initial_time``
+
+        Shifts ``initial_time`` to ``initial_time+value``.
+
+        Args:
+            value: The integer value by which ``initial_time`` is to be shifted.
+        """
+        pass
+
+    @property
+    def final_time(self) -> int | None:
+        """Return the final time of the element"""
+        try:
+            return self.initial_time + self.duration
+        except TypeError:
+            return None
+
+
+class IrInstruction(IrElement):
+    """Pulse IR Instruction
+
+    A Pulse IR instruction represents a ``ScheduleBlock`` instruction, with the addition of
+    an ``initial_time`` property.
+    """
+
+    def __init__(self, instruction: Instruction, initial_time: int | None = None):
+        """Pulse IR instructions
+
+        Args:
+            instruction: the Pulse `Instruction` represented by this IR instruction.
+            initial_time (Optional): Starting time of the instruction. Defaults to ``None``
+
+        Raises:
+            PulseError: if ``initial_time`` is not ``None`` and not non-negative integer.
+        """
+        if initial_time is not None and (
+            not isinstance(initial_time, (int, np.integer)) or initial_time < 0
+        ):
+            raise PulseError("initial_time must be a non-negative integer.")
+
+        self._instruction = instruction
+        self._initial_time = initial_time
+
+    @property
+    def instruction(self) -> Instruction:
+        """Return the instruction associated with the IrInstruction"""
+        return self._instruction
+
+    @property
+    def initial_time(self) -> int | None:
+        return self._initial_time
+
+    @property
+    def duration(self) -> int:
+        return self._instruction.duration
+
+    @initial_time.setter
+    def initial_time(self, value: int):
+        """Set ``initial_time``
+
+        Args:
+            value: The integer value of ``initial_time``.
+
+        Raises:
+            PulseError: if ``value`` is not ``None`` and not non-negative integer.
```

Comment:
Setter docstring is [not rendered](https://stackoverflow.com/questions/17478040/how-can-i-generate-documentation-for-a-python-property-setter-using-sphinx). You can move this to property, e.g.

```python
"""A clock time when this instruction is issued.

.. note::
    initial_time value must be either None or non-negative integer.
"""
```